File: src/main/java/com/emirhan/ft_case/controller/CommentController.java
package com.emirhan.ft_case.controller;

import com.emirhan.ft_case.domain.common.BaseResponse;
import com.emirhan.ft_case.domain.common.Status;
import com.emirhan.ft_case.domain.model.Comment;
import com.emirhan.ft_case.domain.service.CommentService;
import com.emirhan.ft_case.domain.util.DateFormatter;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Date;
import java.util.List;

@RestController
@RequiredArgsConstructor
public class CommentController {
    private final CommentService commentService;

    @GetMapping("/comment")
    ResponseEntity<BaseResponse<List<Comment>>> getComments(
            @RequestParam(required = false) Integer userId,
            @RequestParam(required = false) Integer productId,
            @RequestParam(value = "startDate", required = false) String startDateString,
            @RequestParam(value = "endDate", required = false) String endDateString
    ) {
        Date startDate = DateFormatter.parseToDate(startDateString);
        Date endDate = DateFormatter.parseToDate(endDateString);

        boolean isDatesNotNull = startDate != null && endDate != null;

        List<Comment> commentList = null;

        if (isDatesNotNull) {
            if (productId != null) {
                commentList = commentService.getProductCommentsBetweenGivenDates(productId, startDate, endDate);
            } else if (userId != null) {
                commentList = commentService.getUserCommentsBetweenGivenDates(userId, startDate, endDate);
            }
        } else {
            if (productId != null) {
                commentList = commentService.getProductComments(productId);
            } else if (userId != null) {
                commentList = commentService.getUserComments(userId);
            }
        }

        return new ResponseEntity<>(new BaseResponse<>(Status.Success.name, commentList, null), HttpStatus.OK);
    }
}

File: src/main/kotlin/com/emirhan/ft_case/controller/CommentController.kt
package com.emirhan.ft_case.controller

import com.emirhan.ft_case.domain.entity.BaseResponse
import com.emirhan.ft_case.domain.entity.CommentEntity
import com.emirhan.ft_case.domain.entity.Status
import com.emirhan.ft_case.domain.mapper.DateFormatter.toDate
import com.emirhan.ft_case.domain.repository.CommentRepository
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class CommentController(private val commentRepository: CommentRepository) {
    @PostMapping("/comment")
    fun addComment(@RequestBody comment: CommentEntity): ResponseEntity<BaseResponse<Any>> {
        commentRepository.addComment(comment)
        return ResponseEntity(BaseResponse(Status.Success.message), HttpStatus.OK)
    }

    @GetMapping("/comment")
    fun getComments(
        @RequestParam userId: Int?,
        @RequestParam productId: Int?,
        @RequestParam startDate: String?,
        @RequestParam endDate: String?
    ): ResponseEntity<BaseResponse<List<CommentEntity>>> {
        var commentList = listOf<CommentEntity>()
        if (userId != null) {
            commentList = if (startDate != null && endDate != null) {
                commentRepository.getUserCommentsBetweenDates(userId, startDate.toDate(), endDate.toDate())
            } else {
                commentRepository.getUserComments(userId)
            }
        } else {
            if (productId != null) {
                commentList = if (startDate != null && endDate != null) {
                    commentRepository.getProductCommentsBetweenDates(productId, startDate.toDate(), endDate.toDate())
                } else {
                    commentRepository.getProductComments(productId)
                }
            }
        }

        return ResponseEntity(BaseResponse(Status.Success.message, commentList), HttpStatus.OK)
    }

}